docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,194 @@
+Use `FHU::Filesystem::GetFilename` instead.
+
+#### std::filesystem::copy_file
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
+Use a combination of open and fextl::string APIs instead of fstream.
+
+### `std::fwrite`
+This API allocates a buffer in the background for buffering output.
+Use raw `write` instead.
+
+### `std::string`
+Use `fextl::string` instead.
+
+#### std::to_string
+Use `fextl::fmt::format` instead.
+
+### `std::stol`
+### `std::stoul`
+### `std::stoll`
+### `std::stoull`
+These all consume a `std::string` as their first argument. Use the equivalent functions that don't use `std::string`
+- `std::strtol`
+- `std::strtoul`
+- `std::strtoll`
+- `std::strtoull`
+
+### `fmt::`
+### `fmt::format`
+Use `fextl::fmt::` instead
+
+### APIs that FEX doesn't have a replacement for
+Don't use any of these APIs in FEXLoader/FEXInterpreter. Shoutout to
+[this](https://stackoverflow.com/questions/43056338/standard-library-facilities-which-allocate-but-dont-use-an-allocator) StackOverflow post for this
+huge list.
+
+#### `std::any`
+#### `std::function` and lambdas
+One must take additional considerations when using these to ensure that they don't allocate memory. These don't have any way to replace which
+allocator is used for these objects. Additionally there is no way up-front to know if these will allocate memory or if the compiler will use
+small-function optimizations to avoid allocations. The only real way to check this is to enable the glibc faulting compile option.
+- Lambdas without anything in the capture list will never allocate memory.
+- One pointer in the capture list is likely to hit small-function optimizations and not allocate memory.
+   - This isn't guaranteed.
+- Passing the `std::function` as an argument is unlikely to optimize away their memory allocations.
+
+#### `std::valarray`
+#### `std::filebuf`
+#### `std::inplace_merge`
+#### `<stdexcept>`
+#### `std::boyer_moore_searcher`
+#### `std::filesystem::path`
+#### `std::filesystem::directory_iterator`
+#### `std::regex`
+#### `std::thread`
+#### `std::async`
+#### `std::packaged_task`
+#### `std::promise`
+#### `<iostream>`
+#### Remember this is not an all-encompassing list! We may find APIs that still allocate memory and need to be avoided!
+
+### Regular memory allocation routines.
+Don't use these directly as they will the glibc allocator.
+
+#### mmap
+Use `FEXCore::Allocator::mmap`
+
+#### munmap
+Use `FEXCore::Allocator::munmap`
+
+#### malloc
+Use `FEXCore::Allocator::malloc`
+
+#### calloc
+Use `FEXCore::Allocator::calloc`
+
+#### memalign 
+Use `FEXCore::Allocator::memalign`
+
+#### valloc
+Use `FEXCore::Allocator::valloc`
+
+#### posix_memalign
+Use `FEXCore::Allocator::posix_memalign`
+
+#### realloc
+Use `FEXCore::Allocator::realloc`
+
+#### free
+Use `FEXCore::Allocator::free`
+
+#### aligned_alloc
+Use `FEXCore::Allocator::aligned_alloc`
+
+#### __libc_malloc
+**Don't use**
```

Comment:
This is formatted poorly, since the section nesting depth is so deep that the header gets printed in the same style as the `Don't use` text.

Reply:
Changed.